adagrams/game.py

```diff
@@ -1,11 +1,124 @@
+SCORE_DICT = { 'A': 1, 
+'B': 3, 
+'C': 3, 
+'D': 2, 
+'E': 1, 
+'F': 4, 
+'G': 2,  
+'H': 4, 
+'I': 1, 
+'J': 8, 
+'K': 5, 
+'L': 1, 
+'M': 3, 
+'N': 1, 
+'O': 1, 
+'P': 3, 
+'Q': 10, 
+'R': 1, 
+'S': 1,
+'T': 1, 
+'U': 1, 
+'V': 4, 
+'W': 4, 
+'X': 8, 
+'Y': 4, 
+'Z': 10}
+
 def draw_letters():
-    pass
+    adjusted_letter_pool=LETTER_POOL.copy()
+    hand_list = []
+    
+    #Appends 10 random letters to the hand
+    for _ in range(10):
+        current_letter = random.choice(adjusted_letter_pool)
+        adjusted_letter_pool.remove(current_letter)
+        hand_list.append(current_letter)
+
+    return hand_list
+
+def uppercase_letter_list(letter_bank):
+    hand=[]
+
+    #Converts each letter in the letter bank to uppercase
+    for letter in letter_bank:
+        upper_letter = letter.upper()
+        hand.append(upper_letter)
+
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    players_letters_list = list(word.upper())
+    hand = uppercase_letter_list(letter_bank)
+    hand_dict = {}
+
+    #Generates a counter dictionary for letters in the hand
+    for hand_letter in hand:
+        # lowercase_hand_letter = hand_letter.upper()
+        if hand_letter in hand_dict:
+            hand_dict[hand_letter] += 1
+        else:
+            hand_dict[hand_letter] = 1
 
-def score_word(word):
-    pass
+    #Checks validity of user's word letters against the counter dictionary
+    for player_letter in players_letters_list:
+        if player_letter not in hand_dict:
+            return False
+
+        if hand_dict[player_letter] == 0:
+            return False
+        else:
+            hand_dict[player_letter] -= 1
+    #Assumes a valid word if the function does not return False
+    return True
+
+def score_word(word):    
+    players_letters_list=list(word.upper())
+    letter_count=len(players_letters_list)
+    players_word_score = 0
+
+    #Adds each letter's points to the word score 
+    for player_letter in players_letters_list:
+        players_word_score += SCORE_DICT[player_letter]
+    
+    #Adds 8 points to the word score if the word is 7 letters or longer
+    if letter_count >= 7:
+        players_word_score += 8
+        
+    return players_word_score
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    max_words = []
+    letter_counts = []
+    
+    #assigns an initial max score
+    max_score = score_word(word_list[0])
+    
+    # compares current score against previously stored max score 
+    # and appends (1) the word and (2) number of letters in the word to lists
+    for word in word_list:
+        current_score = score_word(word)
+        if current_score > max_score:
+            max_score = current_score
+            max_words = [word]
+            letter_counts = [len(word)]
+        elif current_score == max_score:
+            max_words.append(word)
+            letter_counts.append(len(word))
+
+    min_letter_count = min(letter_counts)
```

Comment:
Hmm the logic here is a little hard to follow. We have two lists that are keeping track of values that relate to each other. That's a hint that we should use a dictionary to carry the word and its length together as key-value pairs.

I think, however, we can do all of this in one for loop without creating two more below, which iterate over the same list twice. It also gets rid of the two lists `max_words` and `letter_counts`.

Maybe something like:
```python
for word in word_list:
    current_score = score_word(word)
    if current_score > max_score:
        # assign the current word to a max_word variable
        # assign the current score to a max_score variable
    elif current_score == max_score:
        # now bring in the logic inside the two for loops below
        # is the current word 10 letters long and longer than the max_word? 
            # it's now the max_word
        # is the current word shorter than the max_word?
            # it's now the max_word
```